### PR TITLE
conditionally unused parameters can be declared using maybe_unused attribute.

### PR DESCRIPTION
**Pull request by apenn-msft**
conditionally unused parameters can be declared using maybe_unused attribute.

### CODE REVIEW

**Comment by hsutter** [comment]
Editors call: Could you give an example of where a parameter might or might not be used in a TU, that is not an ODR violation? We only know cases involving `if constexpr` in templates. Also, the suggestion should be in a Note because it's a special case, rather than in the Reason for the rule.

**Comment by apenn-msft** [comment]
> Editors call: Could you give an example of where a parameter might or might not be used in a TU, that is not an ODR violation? We only know cases involving `if constexpr` in templates. Also, the suggestion should be in a Note because it's a special case, rather than in the Reason for the rule.

moved to notes and added example

    X* find(set<Blob>& s, const string& str, [[maybe_unused]] Hint h) 
    { 
        #if DEBUG
            // a hint is used only in debug builds
        #endif

        if constexpr (Blob::MaxSize > 1000)
        {
            // a hint is used only if blobs are of a certain size
        }
    }

**Review by ChrisGuzak — APPROVED** [review]

**Comment by hsutter** [comment]
Editors call: Thanks! This is better, but a couple of changes please:
- remove "depending on the translation unit" because we think that may not be actually right (it sounds like it's asking for an ODR problem)
- in the body, remove the `ifdef` and make the example a function template and the `if constexpr` condition a dependent condition

**Comment by apenn-msft** [comment]
> Editors call: Thanks! This is better, but a couple of changes please:
> 
> * remove "depending on the translation unit" because we think that may not be actually right (it sounds like it's asking for an ODR problem)
> * in the body, remove the `ifdef` and make the example a function template and the `if constexpr` condition a dependent condition

the translation unit was in terms of translations where DEBUG is defined. I think #ifdef DEBUG is probably the most common example, but yep, sure I can move to an if constexpr example to keep it pure C++ and avoid precompiler

File: CppCoreGuidelines.md [hunk 1]
```diff
@@ -2846,16 +2846,29 @@ Not possible.
 
 Readability.
 Suppression of unused parameter warnings.
-If parameters are conditionally unused depending on the translation unit, declare them with the `[[maybe_unused]]` attribute.
 
 ##### Example
 
-    X* find(map<Blob>& m, const string& s, Hint);   // once upon a time, a hint was used
+    X* find(set<Blob>& s, const string& str, Hint);   // once upon a time, a hint was used
```
Comment (markyin) [inline CppCoreGuidelines.md:2852]: suggest
`X* find(set<Blob>& ranges, const string& value, Hint);`
It's more readable,

**Comment by hsutter** [comment]
Editors call: Thanks. We don't see the requested changes, were you going to update the PR?

**Comment by apenn-msft** [comment]
> Editors call: Thanks. We don't see the requested changes, were you going to update the PR?

yep, will do, along with @markyin's suggestion. look out for a PR update

**Comment by hsutter** [comment]
Editors call: Great, thanks!

**Pull request merged**